Documentation/Relationship/Relationship.md

```diff
@@ -0,0 +1,598 @@
+# Retrieving Relationship Information
+
+This proposal describes how relationship information can be generated from the DSL, introduces some concepts
+to give hints for relationship retrieval and adds mechanisms to manually define hints.
+
+* A **REST** Interface Exporter will use such relationship information to generate Hypermedia information.
+Specifically it will generate hyperlinks which SHOULD be placed in a designated `_links` section.
+A REST exporter MAY also inline those links if it can ensure that no name collisions occur.  
+This document proposes that [RFC 6570 URI Template](https://tools.ietf.org/html/rfc6570) SHOULD be used to encode those links.
+Alternatively there are also more extensive JSON schemas which allow to provide more information about the structure of
+the endpoint (e.g. [JSON API](https://jsonapi.org),
+[JSON Hyper-Schema](https://json-schema.org/draft/2019-09/json-schema-hypermedia.html),
+[Siren](https://github.com/kevinswiber/siren), [HAL](http://stateless.co/hal_specification.html)).
+
+* A **GraphQL** exporter will use such relationship information - additionally to the structure given by `PathComponent`s -
+to generate the query schema.
+
+The following chapters explain how such relationship information is retrieved.
+
+## 1. Using the DSL "structure"
+
+The path structure of the Webservice can be used to infer relationship information.  
+
+Given the following snippet of an example webservice:
+```swift
+@PathParameter var userId: User.ID
+@PathParameter var postId: Post.ID
+
+var content: some Component {
+  Group("user", $userId) {
+    UserHandler(userId: $userId)
+    Group("post", $postId) {
+      PostHandler(userId: $userId, postId: $postId)
+    }
+    Group("static-info") {
+      StaticInfoHandler(userId: $userId)
+      Group("detailed") {
+        DetailedStaticInfoHandler(userId: $userId)
+      }
+    }
+  }
+}
+```
+
+Apodini will create a Relationship for each Endpoint linking to the **first** Endpoint lying deeper in the path, 
```

Comment:
> I would suggest to rephrase this in a way that is states that a RESTful exporter will link to all endpoints that are registered for a URI path with one additional path component.

Exactly that is not the case. It not just selects the endpoint "[...] with **one additional** path component".
The idea was that to handle cases like this:
```swift
var content: some Component = {
  TestHandler()
  Group("path", "to") {
    Test2Handler()
  }
}
```

With your approach the `/` endpoint (the `TestHandler`) would link to `/path`. This is not really useful, as the user didn't specify a Handler there, so it would like to pretty much nothing.

My idea was that for every sub route (in this case the direct and only sub route is  `"path"`), we search the "sub path tree" for the "first" endpoint we encounter, which is `Test2Handler`.

But yes I agree that we probably need to improve phrasing.

Reply:
I see, so so summarize our tactic I could say "It is linking to the first `Component` layer down the component tree that is (1.) separated by at least one path component and (2.) that contains any `Handler`s in the layer"?
As you said:
> for the "first" endpoint we encounter, which is Test2Handler.
Just to be sure I understand "first" in this context, how would we approach the following example.
```
var content: some Component = {
  TestHandler()
  Group("path", "to") {
    Test2Handler() // Uses HTTP GET
    Test3Handler() // Uses HTTP POST
  }
}
```
Would we link to both endpoints at /path/to or just to the first one?

If I understood this correctly we wouldn't link to `Test3Handler` in this example?
```
var content: some Component = {
  TestHandler()
  Group("path", "to") {
    Test2Handler() // Uses HTTP GET
    Test3Handler() // Uses HTTP POST
  }
}
```

Reply:
Your example tackles a pretty different problem the proposal doesn't yet clearly specify how to handle that.
Maybe the issues is also in the current wording of the proposal. A relationship (currently) doesn't care about the `Operation`. And I think that's the right way to do it. After all the hyperlink will just be `/path/to` (we might then want to return in e.g. a `OPTIONS` or `INFO` route what different endpoints are available at that route). This approach also fits well with GraphQL I think. The relationship will exist anyways, and depending on the `Operation` a query or mutation will be possible or not.

I think a more interesting example is the following

```swift
var content: some Component = {
  TestHandler()
  Group("path") {
    Group("to0") {
      Test2Handler()
    }
    Group("to1") {
      Test3Handler()
    }
    Group("to2") {
      Group("to2.1") {
        Test4Handler()
      }    
    }
  }
}
```

How I envision it `TestHandler` would link to all three: `/path/to0`, `/path/to1` and `/path/to2/to2.1`.
First of all, I don't think it would be great to make a decision about order (like to take `Test2Handler` because its written few lines above).
Excluding anyone of them would result in those parts not being discoverable. HTTP wouldn't return links to those endpoints and for GraphQL does parts wouldn't even be accessible.
Would love to hear your opinion on that specific example, if you see it the same way as I?

So yes, the phrasing of this part of the proposal most definitely needs some clarifications (and maybe some more detailed examples) 👍 

So the retrieval algorithm looks something like the following: We search through every sub route of the Endpoint, and should we encounter a path which doesn't hold any endpoints we continue to recursively search all of sub routes for that path the same way.
Code description can be found in the experimental rest branch: https://github.com/Apodini/Apodini/blob/45023b3f2c871d7df14f269ec63e9a49f18f9a39/Sources/Apodini/Semantic%20Model%20Builder/EndpointsTree.swift#L172-L181

Reply:
Thanks for the clarifications and the example! I think your approach makes sense and I can follow the algorithm. A agree with the tactic of linking to `/path/to0`, `/path/to1` and `/path/to2/to2.1`.

I think it makes sense that different operation types for RESTful APIs are not yet part of the proposal. Nevertheless I could envision that it might be useful to extend the functionality in the future. to e.g. not only like to ignore the operation type but to link to different operations in a different way. Maybe this would be part of the configuration DSL of a return type of a `Handler`. E.g. I could envision the idea of linking to a route that can delete the current retrieved type, modify the type and so on. But I think the first iteration is already a great step in the right direction. I am confident that the ideas is clearly communicated when you clarify the algorithm as described in the comment above. Tackling different operation types would then be something that we could tackle once we got basic relationships working.

Reply:
The phrasing was addressed with the latest changes. Additional a pseudo code implementation was provided, because I think nothing can describe recursive algorithms better than actual code.
See https://github.com/Apodini/Apodini/pull/41#issuecomment-748679123.
I agree that it will be beneficial to revisit that part of the proposal once we have basic relationships working 👍